jupyterhub_moss/spawner.py

```diff
@@ -224,192 +226,126 @@ async def _get_partitions_info(self):
         resources_info = self.slurm_info_resources(out)
         self.log.debug("Slurm partition resources: %s", resources_info)
 
+        partitions = PartitionsTrait.parse_obj(self.partitions)
+
         # use data from Slurm as base and overwrite with manual configuration settings
         partitions_info = {
-            partition: {**resources_info[partition], **config_partition_info}
-            for partition, config_partition_info in self.partitions.items()
+            partition: PartitionInfo.parse_obj(
+                {
+                    **resources_info[partition].dict(),
+                    **config_partition_info.dict(exclude_none=True),
+                }
+            )
+            for partition, config_partition_info in partitions.items()
         }
-
-        for partition, info in partitions_info.items():
-            for key in REQUIRED_RESOURCES_COUNTS:
-                if key not in info:
-                    raise KeyError(
-                        f"Missing required resource '{key}' for partition '{partition}'"
-                    )
-
-        # Ensure returning a dict that can be modified by the callers
-        return deepcopy(partitions_info)
+        return partitions_info
 
     @staticmethod
-    async def create_options_form(spawner):
+    async def create_options_form(spawner: MOSlurmSpawner) -> str:
         """Create a form for the user to choose the configuration for the SLURM job"""
         partitions_info = await spawner._get_partitions_info()
 
         simple_partitions = [
-            partition for partition, info in partitions_info.items() if info["simple"]
+            partition for partition, info in partitions_info.items() if info.simple
         ]
         if not simple_partitions:
             raise RuntimeError("No 'simple' partition defined: No default partition")
         default_partition = simple_partitions[0]
 
         # Strip prologue from partitions_info:
         # it is not useful and can cause some parsing issues
-        for partition_info in partitions_info.values():
-            for env_info in partition_info["jupyter_environments"].values():
-                env_info.pop("prologue", None)
+        partitions_dict = {
+            name: info.dict(
+                exclude={
+                    "jupyter_environments": {
+                        env_name: {"prologue"} for env_name in info.jupyter_environments
+                    }
+                }
```

Comment:
Using pydantic `exclude`... but there looks to be no way to say exclude from all keys in a dict (only all entry in a list/tuple).

Reply:
You could configure the field `prologue` through `Field` to exclude itself from the `dict` representation: https://docs.pydantic.dev/usage/exporting_models/#model-and-field-level-include-and-exclude